File: iosApp/iosApp/Modules/NFTsDashboard/NFTsDashboardViewModel.swift
// Created by web3d4v on 24/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

enum NFTsDashboardViewModel {
    
    case loading
    case error(EmbeddedErrorCollectionViewCell.ViewModel)
    case loaded(nfts: [NFT], collections: [Collection])
}

extension NFTsDashboardViewModel {
    
    struct NFT {
        
        let identifier: String
        let image: String
    }
    
    struct Collection {
        
        let identifier: String
        let coverImage: String
        let title: String
        let author: String
    }
}

extension NFTsDashboardViewModel {
    
    var nfts: [NFT] {
        
        switch self {
        case .loading, .error:
            return []
        case let .loaded(nfts, _):
            return nfts
        }
    }

    var collections: [Collection] {
        
        switch self {
        case .loading, .error:
            return []
        case let .loaded(_, collections):
            return collections
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftsDashboard/NFTsDashboardViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.nftsDashboard

import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel

sealed class NFTsDashboardViewModel {
    object Loading: NFTsDashboardViewModel()
    data class Error(val error: ErrorViewModel): NFTsDashboardViewModel()
    data class Loaded(
        val nfts: List<NFT>,
        val collections: List<Collection>,
        val regularCarousel: Boolean
    ): NFTsDashboardViewModel()

    data class NFT(
        val identifier: String,
        val image: String,
        val previewImage: String?,
        val mimeType: String?,
        val fallbackText: String?
    )

    data class Collection(
        val identifier: String,
        val coverImage: String?,
        val title: String,
        val author: String?,
    )
}

fun NFTsDashboardViewModel.nfts(): List<NFTsDashboardViewModel.NFT> {
    return if (this is NFTsDashboardViewModel.Loaded) return this.nfts
    else emptyList()
}

fun NFTsDashboardViewModel.collections(): List<NFTsDashboardViewModel.Collection> {
    return if (this is NFTsDashboardViewModel.Loaded) return this.collections
    else emptyList()
}
